### PR TITLE
Firehose Core and Bigquery Sink Instrumentation

### PR DESCRIPTION
**Pull request by lavkesh**

### CODE REVIEW

File: src/main/java/io/odpf/firehose/metrics/Metrics.java [hunk 1]
```diff
@@ -38,26 +42,67 @@
...
     public static final String ERROR_EVENT = APPLICATION_PREFIX + ERROR_PREFIX + "event";
+    public static final String ERROR_MESSAGES_TOTAL = APPLICATION_PREFIX + ERROR_PREFIX + "messages_total";
 
     // EXECUTION TAGS
     public static final String SUCCESS_TAG = "success=true";
     public static final String FAILURE_TAG = "success=false";
+    public static final String MESSAGE_TYPE_TAG = "type=%s"; // total, success, failure
+    public static final String MESSAGE_SCOPE_TAG = "scope=%s";
+
+    public static final String BIGQUERY_TABLE_TAG = "table=%s";
```
Comment (sravankorumilli) [inline src/main/java/io/odpf/firehose/metrics/Metrics.java:74]: we can scope them properly sink specific constants, Core constants

File: src/main/java/io/odpf/firehose/sink/AbstractSink.java [hunk 2]
```diff
@@ -26,30 +31,56 @@
...
+     * @return the list of failed messages
      * @throws DeserializerException when invalid kafka message is encountered
      */
     public List<Message> pushMessage(List<Message> messages) throws DeserializerException {
-        List<Message> failedMessages;
+        List<Message> failedMessages = messages;
         try {
             instrumentation.logDebug("Preparing {} messages", messages.size());
+            instrumentation.captureMessageBatchSize(messages.size());
+            instrumentation.captureMessageMetrics(Metrics.SINK_MESSAGES_TOTAL, Metrics.MessageType.TOTAL, messages.size());
```
Comment (sravankorumilli) [inline src/main/java/io/odpf/firehose/sink/AbstractSink.java:42]: are message metrics solving anything better, if we have have global metrics that should be sufficient. I believe we might not  need this metric

File: src/main/java/io/odpf/firehose/sink/bigquery/BigQuerySink.java [hunk 3]
```diff
@@ -47,10 +43,12 @@ public BigQuerySink(Instrumentation instrumentation,
     protected List<Message> execute() throws Exception {
         Instant now = Instant.now();
         Records records = converterCache.getMessageRecordConverter().convert(messageList, now);
-        InsertAllResponse response = insertIntoBQ(records.getValidRecords());
         List<Message> invalidMessages = records.getInvalidRecords().stream().map(Record::getMessage).collect(Collectors.toList());
-        if (response.hasErrors()) {
-            invalidMessages.addAll(BigQueryResponseParser.parseResponse(records.getValidRecords(), response));
+        if (records.getValidRecords().size() > 0) {
```
Comment (sravankorumilli) [inline src/main/java/io/odpf/firehose/sink/bigquery/BigQuerySink.java:47]: this is just whitespace can be removed

File: src/main/java/io/odpf/firehose/sinkdecorator/SinkFinal.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package io.odpf.firehose.sinkdecorator;
```
Comment (sravankorumilli) [inline src/main/java/io/odpf/firehose/sinkdecorator/SinkFinal.java:1]: we can just capture the global metric in abstract sink itself, the current responsibility i believe is not good enough to be a seperate class.

File: src/main/java/io/odpf/firehose/sinkdecorator/SinkWithDlq.java [hunk 5]
```diff
@@ -74,25 +79,27 @@ public static SinkWithDlq withInstrumentationFactory(Sink sink, DlqWriter dlqWri
...
-
+        retryQueueMessages.forEach(m -> {
+            if (m.getErrorInfo() == null) {
+                m.setErrorInfo(new ErrorInfo(null, ErrorType.DEFAULT_ERROR));
+            }
+            instrumentation.captureMessageMetrics(DLQ_MESSAGES_TOTAL, Metrics.MessageType.TOTAL, m.getErrorInfo().getErrorType(), 1);
+        });
+        int attemptCount = 1;
+        while (attemptCount <= this.dlqConfig.getDlqMaxRetryAttempts() && !retryQueueMessages.isEmpty()) {
+            instrumentation.incrementCounter(DQL_RETRY_TOTAL);
```
Comment (sravankorumilli) [inline src/main/java/io/odpf/firehose/sinkdecorator/SinkWithDlq.java:90]: typo

File: src/main/java/io/odpf/firehose/metrics/Metrics.java (hunk 1 above)
Comment (lavkesh) [reply src/main/java/io/odpf/firehose/metrics/Metrics.java:74]: yeah. i will put in the different class.

File: src/main/java/io/odpf/firehose/sink/AbstractSink.java (hunk 2 above)
Comment (lavkesh) [reply src/main/java/io/odpf/firehose/sink/AbstractSink.java:42]: Global message metric counts one message only one time... This metric tells how many messages were processed by SINK including retries.

File: src/main/java/io/odpf/firehose/sinkdecorator/SinkFinal.java (hunk 4 above)
Comment (lavkesh) [reply src/main/java/io/odpf/firehose/sinkdecorator/SinkFinal.java:1]: Abstract sink will not have info about ignored messages. It's the bottom most call.. we need to check the messages when everything else is processed.

Comment (lavkesh) [reply src/main/java/io/odpf/firehose/sinkdecorator/SinkFinal.java:1]: consumer-> sinkfinal -> dlq -> retry -> abstractsink
                                                                                 |
            consumer<-sinkfinal <-dlq<-retry<----  V

File: src/main/java/io/odpf/firehose/consumer/GenericConsumer.java [hunk 6]
```diff
@@ -60,6 +61,7 @@ public GenericConsumer(Consumer kafkaConsumer, KafkaConsumerConfig config, Filte
         ConsumerRecords<byte[], byte[]> records = kafkaConsumer.poll(Duration.ofMillis(consumerConfig.getSourceKafkaPollTimeoutMs()));
         instrumentation.logInfo("Pulled {} messages", records.count());
         instrumentation.capturePulledMessageHistogram(records.count());
+        instrumentation.captureGlobalMessageMetrics(Metrics.MessageScope.CONSUMER, records.count());
```
[the remark this answers is not in the document]
Comment (lavkesh) [reply src/main/java/io/odpf/firehose/consumer/GenericConsumer.java:64]: I think it's Fine. Makes it more clear and readable.

**Pull request merged**